src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -0,0 +1,1276 @@
+        return;
+    }
+
+    if( ptr->mainloop )
+    {
+        pa_threaded_mainloop_stop( ptr->mainloop );
+    }
+
+    if( ptr->context )
+    {
+        pa_context_disconnect( ptr->context );
+        pa_context_unref( ptr->context );
+    }
+
+    if( ptr->mainloop )
+    {
+        pa_threaded_mainloop_free( ptr->mainloop );
+    }
+
+    PaUtil_FreeMemory( ptr );
+}
+
+
+void PaPulseAudio_CheckContextStateCb( pa_context * c,
+                                      void *userdata )
+{
+    PaPulseAudio_HostApiRepresentation *ptr = userdata;
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if( !c )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_CheckContextStateCb: Out of memory" );
+        pa_threaded_mainloop_signal(ptr->mainloop, 0);
+        return;
+    }
+
+    ptr->state = pa_context_get_state( c );
+    pa_threaded_mainloop_signal( ptr->mainloop, 0 );
+}
+
+void PaPulseAudio_ServerInfoCb( pa_context *c,
+                                const pa_server_info *i,
+                                void *userdata )
+{
+  PaPulseAudio_HostApiRepresentation *l_ptrHostApi =
+      (PaPulseAudio_HostApiRepresentation *) userdata;
+  PaError result = paNoError;
+  const char *l_strName = NULL;
+
+  l_ptrHostApi->pulseaudioDefaultSource[0] = '\0';
+  l_ptrHostApi->pulseaudioDefaultSink[0] = '\0';
+
+  if( i->default_sink_name != NULL )
+  {
+    /* Make sure there is '\0' is at the end of string with using snprintf */
+    snprintf( l_ptrHostApi->pulseaudioDefaultSink,
+              PAPULSEAUDIO_MAX_DEVICENAME,
+              "%s",
+              i->default_sink_name );
+  }
+
+  if( i->default_source_name != NULL )
+  {
+    /* Make sure there is '\0' is at the end of string with using snprintf */
+    snprintf( l_ptrHostApi->pulseaudioDefaultSource,
+              PAPULSEAUDIO_MAX_DEVICENAME,
+              "%s",
+              i->default_source_name );
+  }
+
+  pa_threaded_mainloop_signal( l_ptrHostApi->mainloop,
+                               0 );
+}
+
+
+int _PaPulseAudio_AddAudioDevice( PaPulseAudio_HostApiRepresentation *hostapi,
+                                  const char *PaPulseAudio_SinkSourceName,
+                                  const char *PaPulseAudio_SinkSourceNameDesc,
+                                  int inputChannels,
+                                  int outputChannels,
+                                  double defaultLowInputLatency,
+                                  double defaultHighInputLatency,
+                                  double defaultLowOutputLatency,
+                                  double defaultHighOutputLatency,
+                                  const long defaultSampleRate )
+{
+    /* These should be at least 1 */
+    int l_iRealNameLen = strnlen(PaPulseAudio_SinkSourceNameDesc, PAPULSEAUDIO_MAX_DEVICENAME) + 1;
```

Comment:
Len means without '\0'
```suggestion
    int l_iRealNameSize = strnlen(PaPulseAudio_SinkSourceNameDesc, PAPULSEAUDIO_MAX_DEVICENAME - 1) + 1;
```
The same below